Update service to ESM and WDIO v9

dons20 opened the pull request:
## Description
This started as a small experiment to make some changes locally to update the service. I figured it may be a good idea to update this to use ESM as well as Node v20 since it's the latest LTS at this time.

Leaving this up as a draft to get some initial thoughts. I'm not particularly opinionated as to the direction this should go either. Will need to do additional testing so that everything is compatible, but hopefully some of these changes help 🙂

## Checklist
- [x] Unit/Integration test added (if applicable)
- [x] Documentation added/updated (wiki or md)
- [x] Code style is consistent with the rest of the project

Commit:
Updated project to ESM only
Updated to WDIO v8
Updated to Node v18

Commit:
Additional config updates
Switch to jsconfig for now
Updated necessary types for spec files
Specify browserVersion for wdio conf

christian-bromann commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "wdio-docker-service",
-  "version": "3.3.1",
+  "version": "4.0.0",
```
This will be automatically done by our release pipeline, so let's revert this

christian-bromann commented on package.json:
```diff
@@ -30,30 +31,32 @@
-    "@wdio/mocha-framework": "^7.30.2",
-    "@wdio/selenium-standalone-service": "^7.30.2",
-    "@wdio/spec-reporter": "^7.30.2",
+    "@babel/cli": "^7.23.0",
+    "@babel/core": "^7.23.2",
+    "@babel/preset-env": "^7.23.2",
+    "@babel/register": "^7.22.15",
```
Can we switch to TypeScript by chance? It would align with how we develop other services too.

dons20 replied:
Definitely. It was something I was considering but wasn't sure if that was the intended direction. Will have it on my todo list

christian-bromann replied:
It seems this will be a bit of a larger rewrite anyway so let's add it right away.

dons20 replied:
Updated now to include typescript in the src files, and removed babel references

christian-bromann commented on src/launcher.js:
```diff
@@ -91,4 +91,5 @@ class DockerLauncher {
 }
 
-module.exports = DockerLauncher;
+export { DockerLauncher };
```
```suggestion
```

christian-bromann commented on test/unit/modules/dockerEventsSpec.js:
```diff
@@ -1,8 +1,8 @@
 import { expect } from 'chai';
 import { stub, spy } from 'sinon';
-import ChildProcess from 'child_process';
+import * as ChildProcess from 'child_process';
```
```suggestion
import ChildProcess from 'child_process';
```

christian-bromann commented on src/utils/dockerEventsListener.js:
```diff
@@ -1,6 +1,8 @@
 import EventEmitter from 'events';
 import { fork } from 'child_process';
 import path from 'path';
+import * as url from 'url';
```
```suggestion
import url from 'url';
```

Commit:
Reverted manual version change
Replaced import all with default import
Removed unnecessary export

christian-bromann commented:
@dons20 thanks for taking a stab at it. Happy to help if needed.

dons20 commented:
Haven't forgotten about this one, I had some major PC issues. SSD failed but I got it replaced. Will have some changes out soon again.

dons20 committed:
Updates to dependencies and tsconfig

dons20 committed:
Updated js src files to ts
Removed older babel references, configs
Added types in place of JSDoc types

dons20 committed:
Removed additional babel references
Update files to correct line-endings

dons20 committed:
More TS Changes

christian-bromann commented:
Hey @dons20 , thanks for all your work so far! Let me know when you think this is ready for review!

dons20 commented:
Will do, I only have to retest the test directory with these new changes. Everything else should be done

dons20 committed:
Updated most of test directory to ts
Updated deps
Added temporary file for further run arg types and comments

dons20 commented:
Back at it, pushed some of my changes, and I'll have some more incoming soon. Trying to get the local build and tests working fully.

christian-bromann commented:
Awesome, looking forward to the changes. Let me know when I can review something.

dons20 committed:
Updated remaining test files with TS
Added mocks where needed
Updated tsconfigs to handle build vs development type-checking
Updated imports where applicable
Other minor improvements

dons20 committed:
WIP change to update tsconfig layout

Commit:
Fixed some issues preventing unit tests to run
Updated more configs
Added some dev dependencies

dons20 commented:
I'm pretty happy with the results so far, and I think most of this can start to be reviewed, but you can let me know if I should take this out of draft yet. I still have these items on my todo list:

- Fixing failing unit tests and verify coverage is displaying correctly
- Fixing WDIO config so that it connects in its integration tests
- Updating CJS directory to output the necessary cjs module (I took inspiration from some other services which implemented this such as wdio-nuxt-service)
- Verifying output from build

christian-bromann commented on src/utils/ping.ts:
```diff
@@ -0,0 +1,9 @@
+import fetch from 'node-fetch';
+
+class PingClass {
```
Are your planning to add anything to this class? If not I guess it makes sense to not abstract a single `fetch` call

dons20 replied:
No longer an issue with the latest updates.

christian-bromann reviewed (commented):
Some comments, overall looks aweseome!

dons20 commented:
Thanks for reviewing! Will go through the comments shortly

Commit:
Address most of the feedback

dons20 commented on src/launcher.ts:
```diff
@@ -0,0 +1,117 @@
+        } = config;
+
+        if (!image) {
+            return Promise.reject(new Error('dockerOptions.image is a required property'));
```
I've updated it to use the `SevereServiceError`. Let me know if that looks good.

dons20 commented on src/launcher.ts:
```diff
@@ -0,0 +1,117 @@
+    }
+}
+
+class DockerLauncherForTests extends DockerLauncher {
```
When writing the tests, it complained about some variables being private or protected when trying to stub or spy. Perhaps there's a better way of resolving this?

dons20 commented on src/utils/docker.ts:
```diff
@@ -0,0 +1,478 @@
+    startDelay?: number;
+};
+
+export type DockerRunArgs = {
```
Not that I'm aware of unfortunately. If a good, updated one exists, I'd be happy to migrate to it.

dons20 commented on src/launcher.ts:
```diff
@@ -0,0 +1,117 @@
+        try {
+            await this.docker.run();
+            if (typeof onDockerReady === 'function') {
+                onDockerReady();
```
Well considering that `run` returns a promise, would there be any disadvantages to awaiting this? Or do you mean that we should only execute the following code after the promise resolves?

Commit:
Update onPrepare to promise syntax instead of async/await

christian-bromann commented:
@dons20 any updates or anything I can help with?

dons20 commented:
No updates yet, had a couple of busy weeks but I will make some updates in the coming days 🤞🏾

seanpoulter commented:
Would it help if I separated out the upgrade from v7 to v8 to fix the CI pipeline added in #132?

dons20 commented:
@seanpoulter If that is something you can manage in a short period of time, then feel free to

seanpoulter commented:
It's probably no surprise but the update from v7 to v8 requires loading the ES Module `@wdio/logger` from the CJS code. I don't know how to do that as a concise change. :facepalm:

Let me know if there's anything else I can do to help.

dons20 commented:
Totally understand. I'm pushing some changes within this weekend, hopefully it will advance this PR much further so it can cover all the concerns.

Commit:
Merge https://github.com/webdriverio-community/wdio-docker-service into ESM-update

Commit:
Some fixes to failing tests
More usage of sandbox instead of direct stubs

Commit:
Fixed tests in launcherSpec
Removed console log
Fixed some warnings and errors displayed in some test contexts

dons20 commented:
So recently I've been working on the unit tests as they were tricky to get started. After much debugging I had realized that one file was blocking the others from executing. Once I commented some of those tests, the others started running and I was able to fix the failing ones. There are about 20 to resolve on the unit tests side.

I'll continue chipping away at it in the coming days whenever I can find time. Just keeping everyone in the loop.

Commit:
Fixed dockerEventsSpec

dons20 committed:
Fixed all test cases
Updated to latest dependencies

dons20 committed:
Merge remote-tracking branch 'Baseline/main' into ESM-update

dons20 committed:
Updated package-lock

dons20 commented:
Hey everyone, sorry for the delay. The last few months have been quite busy with personal stuff, and there were many test cases which were causing weird / vague errors that I had difficulty in troubleshooting.

Thankfully I was able to finish troubleshooting them with the help of Gemini 3 😁. All of the tests should now be working as expected. I've also bumped all the dependencies to their latest. So hopefully we can begin the review process soon.

dons20 commented on src/cjs/index.ts:
```diff
@@ -0,0 +1,3 @@
+exports.launcher = class DockerLauncher {
+    // WIP
+};
```
What would we need to export here?

dons20 committed:
* Updated CJS entry file
* Removed extra md file

christian-bromann commented on package.json:
```diff
@@ -31,33 +45,40 @@
+    "eslint": "^9.39.1",
+    "esmock": "^2.7.3",
+    "globals": "^16.5.0",
+    "mocha": "^11.7.5",
```
Any chance we can migrate to `vitest`? We use it across the whole project and it contains all features so we don't additionally need `chai`, `esmock`, `sinon` etc.

dons20 replied:
Yeah I'll give it a shot, hopefully not too many changes needed to make it compatible. The only thing is that some mocks might behave differently so I'll see what the results are locally

christian-bromann replied:
Thank you! I am sure Cursor or any other AI IDE can be very helpful here 😉

dons20 replied:
I've done the migration to vitest in my last commit. Copilot ftw for making it much quicker. You can take a look when you have the time.

christian-bromann reviewed (commented):
Some requests 😊

dons20 committed:
Migrated to vitest
Updated some test cases to fix type errors
Updated Github CI config
Updated node version to latest v20 lts

dons20 committed:
Fixed Eslint issues
Cleaned up some older references

dons20 committed:
Removed abstraction on Ping file

christian-bromann reviewed (approved):
LGTM 👍 

Amazing work here 👏

dons20 committed:
Updated health check URL to work across platforms

https://github.com/moby/libnetwork/pull/2348#issuecomment-1177610499

dons20 commented:
@christian-bromann I've made some changes that should ensure that `host.docker.internal` is correctly resolved across all platforms, which should allow it to work on the pipeline. Retry when you can.

Edit: Thanks for the addition to the maintainers! I'll re-run and see the latest result.

dons20 committed:
Updated baseUrl for integration config

dons20 committed:
Added CI check to conditionally use different URL for host

dons20 committed:
Troubleshooting~1

dons20 committed:
Troubleshooting~2

dons20 committed:
Troubleshooting~3

dons20 committed:
Troubleshooting~4

dons20 committed:
Troubleshooting~5

dons20 committed:
Troubleshooting~6

dons20 committed:
Troubleshooting~6

dons20 committed:
Troubleshooting~7

dons20 committed:
Troubleshooting~8

dons20 committed:
Troubleshooting~9

dons20 committed:
Troubleshooting~10

dons20 committed:
Troubleshooting~11

dons20 committed:
Troubleshooting~12

dons20 committed:
Troubleshooting~13

Commit:
Troubleshooting~14

Commit:
Troubleshooting~14

Commit:
Troubleshooting~15

Commit:
Cleanup~1

Commit:
Cleanup~2

Commit:
Cleanup~3

Commit:
Cleanup~4

Commit:
Cleanup~5

Commit:
Troubleshooting~16

Commit:
Troubleshooting~17

Commit:
Troubleshooting~18

Commit:
Troubleshooting~19

Commit:
Troubleshooting~20

Commit:
Troubleshooting~21

Commit:
Troubleshooting~22

Commit:
Troubleshooting~23

Commit:
Troubleshooting~24

Commit:
Remaining Cleanup~1

dons20 commented:
Okay so after a few hours of troubleshooting the pipeline, I was able to identify a few necessary fixes to the code itself. Here's what I identified and fixed.

1. The pipeline would hang at the end of the test step. This was fixed with a call to `process.unref()` so that Node could move on with the remaining steps.
2. After pulling a fresh image, the container wouldn't automatically startup, so the health checks to the URL would never work. Fixed by changing `return` to `await` in `await this._pullImage()`. I identified that if I added a step before the tests to simply pre-fetch the image I wanted to run in the test, the test would actually pass 100% of the time. So it was simply an issue of the image being fetched but not started when needed.
3. Health check polling logic was a bit incorrect in terms of when it would increment the retry counter, so that was updated, though it may not be critical anymore.
4. I had to add a workaround for the CI environment so that it would use the loopback IP address `127.0.0.1` instead of `host.docker.internal`. Note: It might be possible to use the latter on all platforms with another workaround, but I'm exhausted from all the retries 😆 so maybe that's a future improvement.

@christian-bromann you can take a look at the latest changes as well as the pipeline. This should be good to go now.

christian-bromann reviewed (approved):
Amazing work 👏  Thanks a ton!
